src/util/locking/FileSystemResourceLocker.ts

```diff
@@ -0,0 +1,146 @@
+  rootFilePath?: string;
+  /** The path to the directory where locks will be stored (appended to rootFilePath) */
+  lockDirectory?: string;
+  /** Custom settings concerning retrying locks */
+  attemptSettings?: AttemptSettings;
+}
+
+function isCodedError(err: unknown): err is { code: string } {
+  return typeof err === 'object' && err !== null && 'code' in err;
+}
+
+/**
+ * A resource locker making use of the [proper-lockfile](https://www.npmjs.com/package/proper-lockfile) library.
+ * Note that no locks are kept in memory, thus this is considered thread- and process-safe.
+ *
+ * This **proper-lockfile** library has its own retry mechanism for the operations, since a lock/unlock call will
+ * either resolve successfully or reject immediatly with the causing error. The retry function of the library
+ * however will be ignored and replaced by our own LockUtils' {@link retryFunctionUntil} function.
+ *
+ * This allows us to reuse the AttemptSettings semantics similar to the {@link RedisLocker} class.
+ */
+export class FileSystemResourceLocker implements ResourceLocker, Finalizable {
+  protected readonly logger = getLoggerFor(this);
+  private readonly attemptSettings: Required<AttemptSettings>;
+  /** Folder that stores the locks */
+  private readonly lockFolder: string;
+
+  /**
+   * Create a new FileSystemResourceLocker
+   * @param rootFilePath - The rootPath of the filesystem
+   * @param lockDirectory - The path to the directory where locks will be stored (appended to rootFilePath)
+                            _[default is `/.internal/locks`]_
+   * @param attemptSettings - Custom settings concerning retrying locks
+   */
+  public constructor(args: FileSystemResourceLockerArgs = {}) {
+    const { rootFilePath, lockDirectory, attemptSettings } = args;
+    this.attemptSettings = { ...attemptDefaults, ...attemptSettings };
+    this.lockFolder = joinFilePath(rootFilePath ?? './', lockDirectory ?? '/.internal/locks');
+    ensureDirSync(this.lockFolder);
+  }
+
+  /** Stop retry when this error occurs */
+  private readonly stopOnError = (err: any): boolean => isCodedError(err) && err.code === 'ENOTACQUIRED';
+
+  /**
+   * Utility wrapper function for all un/lock operations. The function reference will
+   * be retried with the current {@link AttemptSettings} until the function resolves
+   * or the {@link stopOnError} condition holds.
+   * @param fn - The function reference to retry until successful or stopped.
+   */
+  private async retryUntilError<T>(fn: () => Promise<any>): Promise<T> {
+    return await retryFunctionUntil<any>(toRetryOnError(this.stopOnError, fn), this.attemptSettings);
+  }
+
+  public async acquire(identifier: ResourceIdentifier): Promise<void> {
+    const { path } = identifier;
+    this.logger.debug(`Acquiring lock for ${path}`);
+    try {
+      const opt = this.generateOptions(identifier, defaultLockOptions);
+      await this.retryUntilError((): any => lock(path, opt));
```

Comment:
Does everything still work if all the `any`s in this class get replaced by `unknown`? So much `any` scares me 😄

Reply:
No more anys in the refactored work ;-p